src/beliefpropagation/beliefpropagation_schedule.jl

```diff
@@ -0,0 +1,51 @@
+abstract type EdgeSequenceAlgorithm end
+
+struct ForestCover <: EdgeSequenceAlgorithm end
+struct Parallel <: EdgeSequenceAlgorithm end
+
+default_edge_sequence_alg() = ForestCover()
+
+default_bp_niters(g::NamedGraph) = is_tree(g) ? 1 : nothing
+function default_bp_niters(g::Union{DataGraph,AbstractITensorNetwork})
```

Comment:
Seems like this could be simplified with:
```julia
@traitfn default_bp_niters(g::AbstractGraph::::(!IsDirected)) = is_tree(g) ? 1 : nothing

@traitfn default_bp_niters(g::AbstractGraph::::IsDirected) = default_bp_niters(undirected_graph(g))
```
or something like that.

Reply:
Also looks like something similar could be used for `edge_sequence`.

Reply:
I could make everything just dispatch on abstract graph types by doing, say:

`default_bp_niters(g::AbstractGraph) = is_tree(undirected_graph(underlying_graph(g))) ? 1 : nothing` ?

And similarly with `edge_sequence` ?

Reply:
Hmm looks like it is defined here: https://github.com/JuliaGraphs/Graphs.jl/blob/cc3052f9e3c7486d006f3864fe42da0d4e03e5d7/src/trees/prufer.jl#L11-L13, does it break for `AbstractDataGraph`?

Reply:
Yeah it appears to break for `AbstractDataGraph`

Reply:
Looks ok to me:
```julia
julia> using Graphs

julia> using NamedGraphs

julia> using DataGraphs

julia> is_tree(DataGraph(grid((4, 1))))
true

julia> is_tree(DataGraph(grid((4, 2))))
false

julia> is_tree(DataGraph(named_grid((4, 1))))
true

julia> is_tree(DataGraph(named_grid((4, 2))))
false
```

Reply:
See 
![image](https://github.com/mtfishman/ITensorNetworks.jl/assets/51231103/5e9f6c34-f6a1-4541-8837-e757009d9221)

Reply:
Right well that's just that you are inputting a directed graph, which isn't inherently an issue with calling `is_tree` on a `DataGraph`, just that it's not defined for directed graphs. I'm being pedantic on purpose to try to hone in on what the actual issue is, i.e. `is_tree` is working for _undirected_ `DataGraph`s.

So then I understand why it probably makes sense to use `is_tree(undirected_graph(underlying_graph(g)))` since you want to make it undirected, but it probably doesn't make sense to use `undirected_graph(g)` on a directed graph, since you wouldn't know what to do with the edge data if there are edges in both directions shared between two vertices.

Reply:
Yeah I see what you mean. So now I have made it so all versions of the function which take `g::AbstractGraph` type do `undirected_graph(underlying_graph(g))` and then pass to the version of the which requires an`undirected graph` (which I have made a `traitfn`)

Reply:
Seems like there might be a better pattern for this kind of thing, like defining `undirected_is_tree` and/or `undirected_underlying_graph(g))`. But this seems like a good generic solution for now.

Reply:
Yeah agreed it would be optimal to handle such things in `NamedGraphs`. Thanks for making the issue.